distributed/shuffle/shuffle-design.md

```diff
@@ -0,0 +1,139 @@
+
+Additionally, cleanly stopping a running shuffle takes more than just releasing the shuffle tasks from memory: since there's out-of-band processing going on, the `ShuffleExtension` has to be informed in some way that it needs to stop doing whatever it's doing in the background, and clear out its buffers. Also, executing tasks may be blocking on the `ShuffleExtension` doing something; without a way to tell the extension to shut down, those tasks might block forever, deadlocking the cluster.
+
+Therefore, we propose adding a `RerunGroup` (`ImpureGroup`? `CoExecutionGroup`? `RestartGroup`? `OutOfBandGroup`? name TBD) structure to the scheduler which intertwines the fates of all tasks within it: if any one task is to be rescheduled (due to its worker leaving), all tasks are restarted; if any one is to be prematurely released (due to cancellation), all are released.
+
+Membership in a `RerunGroup` is implemented via task annotations, where each task gives the name of the `RerunGroup` it belongs to. A task can belong to at most one `RerunGroup`. TBD if we will enforce any structural restrictions on `RerunGroup`s to prevent odd/invalid states from emerging—we probably should, such as not allowing disjoint parts of the graph to be in the same `RerunGroup`, etc.
+
+Additionally, the scheduler informs workers whenever a `RerunGroup` is restarted or cancelled. Workers will have a way to pass this information on to any interested out-of-band operations. This could be something like:
+- Workers have a named `threading.Event` for each `RerunGroup` that any of their current tasks belong to. When the scheduler tells workers about a restart/cancellation, they `set()` the corresponding event so that some background thread can respond accordingly.
+- A `register_cancellation_handler(rerun_group: str, async_callback: Callable)` method on workers that registers an async function to be run when that group is cancelled/restarted. A potential upside (and potential deadlock) is that the scheduler's `cancel_rerun_group` RPC to workers could block on this callback completing, meaning the scheduler wouldn't treat the `RerunGroup` as successfully cancelled until every callback on every worker succeeded. That could give us some nice synchronization guarantees (which we may or many not actually need) ensuring a shuffle doesn't try to restart while it's also trying to shut down.
+
+### Peer discovery and initialization
+
+Problems to solve:
+* Workers need to all have the same list of peers participating in the shuffle (otherwise data could end up in two different places!)
+* Scheduler needs to be told where to run the `unpack` tasks which bring data back in-band
+
+We'll run a single `shuffle-setup` task before all the transfers to do some initialization.
```

Comment:
> Why does this need to be a single task?

Because it needs to run exactly once. The main thing this task is doing is getting the list of current workers from the scheduler. If we did this in every task, and workers were joining the cluster at the same time, we'd have different lists on different workers and be sending the data to inconsistent places.

There's a way we could do this instead with a scheduler plugin (register the shuffle ID on the scheduler; the first call stores the list of workers and subsequent calls always return the same one) but adding another component to me seems more complicated, since we then need to manage the lifecycle on that state too.